# Table of contents

#### 1- Importing libraries
#### 2- Creating test data frame
#### 3- Finding mixed columns and changing
#### 4- Consistency with products dataframe:
                                01 Locating missing values
                                02 Addressing missing values
                                03 Finding duplicates
                                04 Addressing duplicates
                                05 Addressing high max
#### 5- Consistency with orders dataframe:
                                01 Checking for mixed data types
                                02 Checking for missing values
                                03 Checking for duplicates

#### 6- Exporting dataframes


## 1 Importing libraries

In [41]:
#importing pandas and numpy
import pandas as pd
import numpy as np
import os

## 2 Creating test data frame

In [42]:
#create a dataframe
df_test = pd.DataFrame()

In [43]:
#creating mixed column
df_test['mix'] = ['a', 'b', 1, True]

In [44]:
df_test.head()

mix
0     a
1     b
2     1
3  True

## 3 Finding mixed columns and changing 

In [45]:
#checking for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [46]:
#making data type in mix string
df_test['mix'] = df_test['mix'].astype('str')

## 4 Consistency with products dataframe

In [47]:
#importing dataframe
path=r'/Users/Amaikuru/Desktop/08-06-2022 Instacart Basket Analysis.nosync'

In [48]:
df_prods=pd.read_csv(os.path.join(path, '02 Data', 'Original', 'products.csv'), index_col=False)

### 01 Locating missing values

In [49]:
#counting the missing values in each column
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [50]:
#creating dataframe with only null values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [51]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

### 02 Addressing missing values

In [52]:
#imputing with mean and median
#df['column with missings'].fillna(mean value, inplace=True)
#df['column with missings'].fillna(median value, inplace=True), df_prods.median()
#linear interpolation for time series data

In [53]:
df_prods.shape

(49693, 5)

In [54]:
#creating dataframe with non null values
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [55]:
df_prods_clean.shape

(49677, 5)

In [56]:
#dropping missing values
#df_prods.dropna(subset = [‘product_name’], inplace = True)
#this overwrites dataframe to frame without missing values

### 03 Finding duplicates

In [57]:
#looking for full duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [58]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

### 04 Addressing duplicates

In [59]:
#create dataframe with no duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [60]:
#looking for inconsistencies in summary
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

#### 01 max price is incredibly high, this should be looked into
#### 02 high max value makes standard deviation for prices very high

### 05 Addressing high max

In [61]:
#looking for prices over $100
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 100]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

In [62]:
#dropping products with high max
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] >100, 'prices'] = np.nan

In [63]:
#observing new max
df_prods_clean_no_dups['prices'].max()

25.0

## 5 Consistency with orders dataframe

In [66]:
vars_list=['order_id', 'user_id', 'order_number', 'orders_day_of_week', 'order_hour_of_day', 'days_since_prior_order']

In [67]:
#importing ords data frame
df_ords= pd.read_csv(os.path.join(path, '02 Data', 'Prepared', 'orders_wrangled.csv'), usecols=vars_list)

### 01 Checking for mixed data types

In [68]:
#checking for mixed types

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### There are no columns with mixed data types

### 02 Checking for missing values

In [69]:
#checking for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### There are 206,209 missing values in the days_since_prior_order column

### 03 Checking for duplicates

In [70]:
#checking for duplicates
df_ords_dups = df_ords[df_ords.duplicated()]

In [71]:
df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [75]:
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

#### There are no duplicates 

## 6 Exporting dataframes

In [73]:
#Exporting dataframes
df_ords.to_csv(os.path.join(path, '02 Data','Prepared', 'orders_cleaned.csv'))

In [74]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared', 'products_cleaned.csv'))